File: integer-to-roman.rb
# @param {Integer} num
# @return {String}
def int_to_roman(num)
    ones = Array.[]('I', 'X', 'C', 'M')
    fives = Array.[]('V', 'L', 'D')
    stk = Array.new
    digit = -1
    while num > 0
        stk.push(num%10)
        num/=10
        digit+=1
    end
    lat = ""
    while stk.size > 0 
        d = stk.pop
        if d == 4 
           lat += ones[digit]
           lat += fives[digit]
        elsif d == 9 
           lat += ones[digit]
           lat += ones[digit+1]
        elsif d >= 1
            base = d >= 5 ? 5 : 1
            lat += base == 1?ones[digit]:fives[digit]
            for i in 1..d-base
                lat += ones[digit]
            end
        end
        digit-=1
    end
    return lat 
end

File: integer-to-roman.java
class Solution {
    public String intToRoman(int num) {
        char[] one = new char[]{'I', 'X', 'C', 'M'};
        char[] five = new char[]{'V', 'L', 'D'};
        Stack<Integer> stk= new Stack<>();
        int digit = -1;
        while (num > 0) {
            stk.push(num%10);
            num/=10;
            digit++;
        }
        StringBuilder sb = new StringBuilder();
        while(!stk.empty()) {
            int d = stk.pop();
            if (d == 4) {
               sb.append(one[digit]);
               sb.append(five[digit]); 
            } else if (d == 9) {
               sb.append(one[digit]);
               sb.append(one[digit+1]); 
            } else if (d >= 1) {
                int base = d >= 5 ? 5 : 1;
                sb.append(base == 1?one[digit]:five[digit]);
                for(int i=d-base;i>0;i--)sb.append(one[digit]);
            }
            digit--;
        }
        return sb.toString();
    }
}
